mountpoint-s3/src/cli.rs

```diff
@@ -448,12 +446,24 @@ impl CliArgs {
         }
     }
 
-    fn prefix(&self) -> Prefix {
-        self.prefix.as_ref().cloned().unwrap_or_default()
-    }
-
-    pub fn s3_path(&self) -> S3Path {
-        S3Path::new(self.bucket_name.to_owned(), self.prefix())
+    pub fn s3_path(&self) -> anyhow::Result<S3Path> {
+        let prefix = self.prefix.as_ref().cloned().unwrap_or_default();
+        if self.bucket_name.starts_with("s3://") {
+            if prefix.as_str().is_empty() {
+                S3Path::from_s3_uri(self.bucket_name.to_owned()).map_err(|e| anyhow!(e))
+            } else {
+                Err(anyhow!("explicit prefix option not allowed with S3 URI"))
+            }
+        } else {
+            S3Path::from_bucket_prefix(self.bucket_name.to_owned(), prefix.clone()).map_err(|e| anyhow!(e))
+        }
+        .map_err(|e| {
+            e.context(format!(
+                "could not parse `BUCKET_NAME` and `prefix`. Found bucket name: '{}' and prefix: '{}'",
```

Comment:
Could this be confusing if the customer is using S3 URIs? Maybe we should have separate context in each block (S3 URI vs S3 Bucket/Prefix)